ibis/pandas/execution/tests/test_localcontext.py

```diff
@@ -0,0 +1,119 @@
+import pandas as pd
+import pandas.util.testing as tm
+import pytest
+
+import ibis
+import ibis.expr.operations as ops
+from ibis.pandas.client import PandasClient, PandasTable
+from ibis.pandas.dispatch import (
+    compute_local_context,
+    is_computable_input,
+    pre_execute,
+)
+
+pytestmark = pytest.mark.pandas
+
+
+def test_execute_with_localcontext(t):
+    # define a pre_execute function that trim data accroding to localcontext
+    @pre_execute.register(PandasTable, PandasClient)
+    def execute_table_with_localcontext(
+        op, client, scope=None, localcontext=None, **kwargs
+    ):
+        # retreive df from dict
+        df = client.dictionary[op.name]
+        try:
+            begin, end = map(pd.to_datetime, localcontext)
+        except ValueError:
+            return {op: df}
+        # filter time col with time context
+        time_col = "plain_datetimes_naive"
+        return {op: df[df[time_col] >= begin][df[time_col] < end]}
+
+    expr = t
+    # define a time context for time-series data
+    context = ("20170101", "20170103")
+
+    # without time context, execute produces every row
+    df_all = expr.execute()
+    assert len(df_all["plain_datetimes_naive"]) == 3
+
+    # with context set, execute produces only rows within context
+    df_within_context = expr.execute(localcontext=context)
+    assert len(df_within_context["plain_datetimes_naive"]) == 1
+
+    del pre_execute.funcs[(PandasTable, PandasClient)]
+    pre_execute.reorder()
+    pre_execute._cache.clear()
+
+
+merge_asof_minversion = pytest.mark.skipif(
+    pd.__version__ < '0.19.2',
+    reason="at least pandas-0.19.2 required for merge_asof",
+)
+
+
+@merge_asof_minversion
+def test_context_adjustment(
+    time_keyed_left, time_keyed_right, time_keyed_df1, time_keyed_df2
+):
+    # pre_execute for trimming table data
+    @pre_execute.register(PandasTable, PandasClient)
+    def execute_table_with_localcontext(
+        op, client, scope=None, localcontext=None, **kwargs
+    ):
+        df = client.dictionary[op.name]
+        try:
+            begin, end = map(pd.to_datetime, localcontext)
+        except ValueError:
+            return {op: df}
+        time_col = "time"
+        return {op: df[df[time_col] >= begin][df[time_col] < end]}
+
+    # define a custom context adjustment rule for asof_join
+    @compute_local_context.register(ops.AsOfJoin, PandasClient)
```

Comment:
I think you can implement this outside the test as an example of the application of this.

Reply:
I think we should also add documentations somewhere explaining how localcontext are used.

* execute_node of leaf node can uses localcontext to trim data, or to pass it as a filter in the database query.
* Node that needs to trim data in post_execute
* If pre_execute preloads data, it should use localcontext to trim data as well.

Reply:
We should also document that in order to use this feature, tables need to have a 'time' column.

Reply:
+1